qdrant-landing/content/articles/scalar-quantization.md

```diff
@@ -0,0 +1,291 @@
+
+![A distribution of the vector values](/articles_data/scalar-quantization/float32-distribution.png)
+
+Our example shows that 99% of the values come from a `[-2.0, 5.0]` range. And the 
+conversion to `int8` will surely lose some precision, so we rather prefer keeping the 
+representation accuracy within the range of 99% of the most probable values and ignoring
+the precision of the outliers. There might be a different choice of the range width, 
+actually, any value from a range `[0, 1]`, where `0` means empty range, and `1` would 
+keep all the values. That's a hyperparameter of the procedure called `quantile`. A value 
+of `0.95` or `0.99` is typically a reasonable choice, but in general `quantile ∈ [0, 1]`.
+
+#### Conversion to integers
+
+Let's talk about conversion to `int8`. Integers also have a finite set of values that
+might be represented. Within a single byte they may represent up to 256 different values,
+either from `[-128, 127]` or `[0, 255]`.
+
+![Value ranges represented by int8](/articles_data/scalar-quantization/int8-value-range.png)
+
+Since we put some boundaries on the numbers that might be represented by the `f32`, and
+`i8` has some natural boundaries, the process of converting the values between those
+two ranges is quite natural:
+
+$$ f32 = \alpha \times i8 + offset $$
+
+$$ i8 = \frac{f32 - offset}{\alpha} $$
+
+The parameters $ \alpha $ and $ offset $ has to be calculated for a given set of vectors, 
+but that comes easily by putting the minimum and maximum of the represented range for 
+both `f32` and `i8`. 
+
+![Float32 to int8 conversion](/articles_data/scalar-quantization/float32-to-int8-conversion.png)
+
+For the unsigned `int8` it will go as following:
+
+$$ \begin{equation}
+\begin{cases} -2 = \alpha \times 0 + offset \\\\ 5 = \alpha \times 255 + offset \end{cases} 
+\end{equation} $$
+
+In case of signed `int8`, we'll just change the represented range boundaries:
+
+$$ \begin{equation}
+\begin{cases} -2 = \alpha \times (-128) + offset \\\\ 5 = \alpha \times 127 + offset \end{cases} 
+\end{equation} $$
+
+For any set of vector values we can simply calculate the $ \alpha $ and $ offset $ and 
+those values have to be stored along with the collection to enable to conversion between
+the types. 
+
+#### Distance calculation
+
+We do not store the vectors in the collections represented by `int8` instead of `float32` 
+just for the sake of compressing the memory. But the coordinates are being used while we 
+calculate the distance between the vectors. Both dot product and cosine distance requires 
+multiplying the corresponding coordinates of two vectors, so that's the operation we 
+perform quite often on `float32`. Here is how it would look like if we perform the 
+conversion to `int8`:
+
+$$ f32 \times f32' = $$
+$$ = (\alpha \times i8 + offset) \times (\alpha \times i8' + offset) = $$
+$$ = \alpha^{2} \times i8 \times i8' + offset \times \alpha \times i8' + offset \times \alpha \times i8 + offset^{2} $$
```

Comment:
```suggestion
$$ = \alpha^{2} \times i8 \times i8' + \underbrace{offset \times \alpha \times i8' + offset \times \alpha \times i8 + offset^{2}}_\text{pre-compute} $$```

It should look like this:
![image](https://user-images.githubusercontent.com/1935623/226982084-bc9e38ec-260f-4f85-a064-0659fcc7dbb9.png)

(tried to highlight individual terms as well, but MathJax could not handle this for some reason =( )